satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,474 @@
+        # NB: 0 may be below the surface... could be a problem for high
+        # resolution imagery in mountainous or high elevation terrain
+        # NB: how tolerant of xarray & dask is this?
+        cth_dataset = resample_dataset(
+                cth_dataset, self.base_area, resampler="nearest",
+                radius_of_influence=50000)
+        (pixel_lon, pixel_lat) = cth_dataset.attrs["area"].get_lonlats(chunks=1024)
+        cth_dataset = cth_dataset.where(np.isfinite(pixel_lon) & np.isfinite(pixel_lat))
+        cth_dataset = cth_dataset.where(cth_dataset.notnull(), 0)
+        return cth_dataset
+
+    def _check_overlap(self, cth_dataset):
+        """Ensure cth_dataset is usable for parallax correction.
+
+        Checks the coverage of ``cth_dataset`` compared to the ``base_area``.  If
+        the entirety of ``base_area`` is covered by ``cth_dataset``, do
+        nothing.  If only part of ``base_area`` is covered by ``cth_dataset``,
+        raise a `IncompleteHeightWarning`.  If none of ``base_area`` is covered
+        by ``cth_dataset``, raise a `MissingHeightError`.
+        """
+        warnings.warn(
+            "Overlap checking not impelemented. Waiting for "
+            "fix for https://github.com/pytroll/pyresample/issues/329")
+
+    def _get_corrected_lon_lat(self, base_lon, base_lat, shifted_area):
+        """Calculate the corrected lon/lat based from the shifted area.
+
+        After calculating the shifted area based on :func:`forward_parallax`,
+        we invert the parallax error and estimate where those pixels came from.
+        For details on the algorithm, see the class docstring.
+        """
+        (shifted_lon, shifted_lat) = shifted_area.get_lonlats(chunks=1024)
+        lon_diff = shifted_lon - base_lon
+        lat_diff = shifted_lat - base_lat
+        # We use the bucket resampler here, because parallax correction
+        # inevitably means there will be 2 source pixels ending up in the same
+        # destination pixel.  We want to choose the biggest shift (max abs in
+        # lat_diff and lon_diff), because the biggest shift corresponds to the
+        # highest clouds, and if we move a 10 km cloud over a 2 km one, we
+        # should retain the 10 km.
+        #
+        # some things to keep in mind:
+        # - even with a constant cloud height, 3 source pixels may end up in
+        #   the same destination pixel, because pixels get larger in the
+        #   direction of the satellite.  This means clouds may shrink as they
+        #   approach the satellite.
+        # - the x-shift is a function of y and the y-shift is a function of x,
+        #   so a cloud that was rectangular at the start may no longer be
+        #   rectangular at the end
+        bur = BucketResampler(self.base_area,
+                              da.array(shifted_lon), da.array(shifted_lat))
+        inv_lat_diff = bur.get_abs_max(lat_diff)
+        inv_lon_diff = bur.get_abs_max(lon_diff)
```

Comment:
Would a future optimization be to update BucketResampler to take multiple DataArrays as input for a single function? Or do we know that with dask the calculations will be shared?

Reply:
That would be a good possible optimisation in the `BucketResampler`, but tests appear to show calculations are already shared in dask.

Reply:
Sounds good.